README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> NATS messaging components for Pip.Services in Go

This library is a part of [Pip.Services](https://github.com/pip-services/pip-services) project.
The Nats module contains a set of components for working with the message queue via NATS server [https://nats.io/](https://nats.io/).

The module contains the following packages:
- [**Build**](https://godoc.org/github.com/pip-services3-go/pip-services3-nats-go/build) - Factory for constructing module components
- [**Connect**](https://godoc.org/github.com/pip-services3-go/pip-services3-nats-go/connect) - Components for creating and configuring a connection with Nats
- [**Queues**](https://godoc.org/github.com/pip-services3-go/pip-services3-nats-go/queues) - Message Queuing components that implement the standard [Messaging](https://github.com/pip-services3-go/pip-services3-messaging-go) module interface

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://godoc.org/github.com/pip-services3-go/pip-services3-nats-go/)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-go/pip-services3-nats-go@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.12+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The Golang version of Pip.Services is created and maintained by:
- **Sergey Seroukhov**

The documentation is written by:
- **Levichev Dmitry**


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Nats specific components for Golang Changelog

## <a name="1.0.5"></a> 1.0.5 (2023-01-12)

- Update dependencies
## <a name="1.0.0"></a> 1.0.0 (2020-06-09)

### Features
* **build** Factory for constructing module components
* **connect** Components for creating and configuring a connection
* **queues** Message Queuing components

test/queues/MessageQueueFixture.go
package test_queues

import (
	"testing"
	"time"

	"github.com/pip-services3-go/pip-services3-messaging-go/queues"
	"github.com/stretchr/testify/assert"
)

type MessageQueueFixture struct {
	queue queues.IMessageQueue
}

func NewMessageQueueFixture(queue queues.IMessageQueue) *MessageQueueFixture {
	c := MessageQueueFixture{
		queue: queue,
	}
	return &c
}

func (c *MessageQueueFixture) TestSendReceiveMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// if c.queue.GetCapabilities().CanMessageCount() {
	// 	count, rdErr := c.queue.MessageCount()
	// 	assert.Nil(t, rdErr)
	// 	assert.Greater(t, count, (int64)(0))
	// }

	envelope2, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestMessageCount(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	c.queue.Send("", envelope1)

	time.Sleep(500 * time.Millisecond)

	count, err := c.queue.ReadMessageCount()
	assert.Nil(t, err)
	assert.True(t, count >= 1)
}

func (c *MessageQueueFixture) TestReceiveSendMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))

	time.AfterFunc(500*time.Millisecond, func() {
		c.queue.Send("", envelope1)
	})

	envelope2, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestReceiveCompleteMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// count, rdErr := c.queue.ReadMessageCount()
	// assert.Nil(t, rdErr)
	// assert.Greater(t, count, (int64)(0))

	envelope2, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	cplErr := c.queue.Complete(envelope2)
	assert.Nil(t, cplErr)
	assert.Nil(t, envelope2.GetReference())
}

func (c *MessageQueueFixture) TestReceiveAbandonMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	abdErr := c.queue.Abandon(envelope2)
	assert.Nil(t, abdErr)

	envelope2, rcvErr = c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestSendPeekMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	envelope2, pkErr := c.queue.Peek("")
	assert.Nil(t, pkErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	// pop message from queue for next test
	_, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
}

func (c *MessageQueueFixture) TestPeekNoMessage(t *testing.T) {
	envelope, pkErr := c.queue.Peek("")
	assert.Nil(t, pkErr)
	assert.Nil(t, envelope)
}

func (c *MessageQueueFixture) TestMoveToDeadMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive("", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	mvErr := c.queue.MoveToDeadLetter(envelope2)
	assert.Nil(t, mvErr)
}

func (c *MessageQueueFixture) TestOnMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	receiver := &TestMsgReceiver{}
	c.queue.BeginListen("", receiver)

	time.Sleep(1000 * time.Millisecond)

	sndErr := c.queue.Send("", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(1000 * time.Millisecond)

	envelope2 := receiver.envelope
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	c.queue.EndListen("")
}

type TestMsgReceiver struct {
	envelope *queues.MessageEnvelope
}

func (c *TestMsgReceiver) ReceiveMessage(envelope *queues.MessageEnvelope, queue queues.IMessageQueue) (err error) {
	c.envelope = envelope
	return nil
}


test/connect/NatsConnection_test.go
package test_connect

import (
	"os"
	"testing"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	connect "github.com/pip-services3-go/pip-services3-nats-go/connect"
	"github.com/stretchr/testify/assert"
)

type natsConnectionTest struct {
	connection *connect.NatsConnection
}

func newNatsConnectionTest() *natsConnectionTest {
	natsUri := os.Getenv("NATS_SERVICE_URI")
	natsHost := os.Getenv("NATS_SERVICE_HOST")
	if natsHost == "" {
		natsHost = "localhost"
	}

	natsPort := os.Getenv("NATS_SERVICE_PORT")
	if natsPort == "" {
		natsPort = "4222"
	}

	natsToken := os.Getenv("NATS_TOKEN")
	// if natsToken == "" {
	// 	natsToken = ""
	// }
	natsUser := os.Getenv("NATS_USER")
	// if natsUser == "" {
	// 	natsUser = ""
	// }
	natsPassword := os.Getenv("NATS_PASS")
	// if natsPassword == "" {
	// 	natsPassword = ""
	// }

	if natsUri == "" && natsHost == "" {
		return nil
	}

	connection := connect.NewNatsConnection()
	connection.Configure(cconf.NewConfigParamsFromTuples(
		"connection.uri", natsUri,
		"connection.host", natsHost,
		"connection.port", natsPort,
		"credential.token", natsToken,
		"credential.username", natsUser,
		"credential.password", natsPassword,
	))

	return &natsConnectionTest{
		connection: connection,
	}
}

func (c *natsConnectionTest) TestOpenClose(t *testing.T) {
	err := c.connection.Open("")
	assert.Nil(t, err)
	assert.True(t, c.connection.IsOpen())
	assert.NotNil(t, c.connection.GetConnection())

	err = c.connection.Close("")
	assert.Nil(t, err)
	assert.False(t, c.connection.IsOpen())
	assert.Nil(t, c.connection.GetConnection())
}

func TestNatsConnection(t *testing.T) {
	c := newNatsConnectionTest()
	if c == nil {
		return
	}

	t.Run("Open and Close", c.TestOpenClose)
}


test/queues/NatsMessageQueue_test.go
package test_queues

import (
	"os"
	"testing"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	queues "github.com/pip-services3-go/pip-services3-nats-go/queues"
)

type natsMessageQueueTest struct {
	queue   *queues.NatsMessageQueue
	fixture *MessageQueueFixture
}

func newNatsMessageQueueTest() *natsMessageQueueTest {
	natsUri := os.Getenv("NATS_SERVICE_URI")
	natsHost := os.Getenv("NATS_SERVICE_HOST")
	if natsHost == "" {
		natsHost = "localhost"
	}

	natsPort := os.Getenv("NATS_SERVICE_PORT")
	if natsPort == "" {
		natsPort = "4222"
	}

	natsQueue := os.Getenv("NATS_QUEUE")
	if natsQueue == "" {
		natsQueue = "test"
	}

	natsToken := os.Getenv("NATS_TOKEN")
	if natsToken == "" {
		natsToken = ""
	}

	natsUser := os.Getenv("NATS_USER")
	if natsUser == "" {
		natsUser = "nats"
	}
	natsPassword := os.Getenv("NATS_PASS")
	if natsPassword == "" {
		natsPassword = "nats"
	}

	if natsUri == "" && natsHost == "" {
		return nil
	}

	queue := queues.NewNatsMessageQueue(natsQueue)
	queue.Configure(cconf.NewConfigParamsFromTuples(
		"connection.uri", natsUri,
		"connection.host", natsHost,
		"connection.port", natsPort,
		"credential.token", natsToken,
		"credential.username", natsUser,
		"credential.password", natsPassword,
		"options.autosubscribe", true,
	))

	fixture := NewMessageQueueFixture(queue)

	return &natsMessageQueueTest{
		queue:   queue,
		fixture: fixture,
	}
}

func (c *natsMessageQueueTest) setup(t *testing.T) {
	err := c.queue.Open("")
	if err != nil {
		t.Error("Failed to open queue", err)
		return
	}

	// err = c.queue.Clear("")
	// if err != nil {
	// 	t.Error("Failed to clear queue", err)
	// 	return
	// }
}

func (c *natsMessageQueueTest) teardown(t *testing.T) {
	err := c.queue.Close("")
	if err != nil {
		t.Error("Failed to close queue", err)
	}
}

func TestNatsMessageQueue(t *testing.T) {
	c := newNatsMessageQueueTest()
	if c == nil {
		return
	}

	c.setup(t)
	t.Run("Send Receive Message", c.fixture.TestSendReceiveMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Receive Send Message", c.fixture.TestReceiveSendMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Send Peek Message", c.fixture.TestSendPeekMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Peek No Message", c.fixture.TestPeekNoMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("On Message", c.fixture.TestOnMessage)
	c.teardown(t)
}


connect/NatsConnection.go
package connect

import (
	"time"

	"github.com/nats-io/nats.go"
	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	clog "github.com/pip-services3-go/pip-services3-components-go/log"
)

/**
 NATS connection using plain driver.
 By defining a connection and sharing it through multiple message queues
 you can reduce number of used connections.

 ### Configuration parameters ###
- connection(s):
  - discovery_key:             (optional) a key to retrieve the connection from IDiscovery
  - host:                      host name or IP address
  - port:                      port number (default: 27017)
  - uri:                       resource URI or connection string with all parameters in it
- credential(s):
  - store_key:                 (optional) a key to retrieve the credentials from ICredentialStore
  - username:                  user name
  - password:                  user password
- options:
  - retry_connect:        (optional) turns on/off automated reconnect when connection is log (default: true)
  - max_reconnect:        (optional) maximum reconnection attempts (default: 3)
  - reconnect_timeout:    (optional) number of milliseconds to wait on each reconnection attempt (default: 3000)
  - flush_timeout:        (optional) number of milliseconds to wait on flushing messages (default: 3000)

### References ###
 - \*:logger:\*:\*:1.0           (optional) ILogger components to pass log messages
 - \*:discovery:\*:\*:1.0        (optional) IDiscovery services
 - \*:credential-store:\*:\*:1.0 (optional) Credential stores to resolve credentials
*/
type NatsConnection struct {
	defaultConfig *cconf.ConfigParams
	// The logger.
	Logger *clog.CompositeLogger
	// The connection resolver.
	ConnectionResolver *NatsConnectionResolver
	// The configuration options.
	Options *cconf.ConfigParams

	// The NATS connection object.
	Connection *nats.Conn

	// Topic subscriptions
	subscriptions []*NatsSubscription

	retryConnect     bool
	maxReconnect     int
	reconnectTimeout int
	flushTimeout     int
}

// NewNatsConnection creates a new instance of the connection component.
func NewNatsConnection() *NatsConnection {
	c := &NatsConnection{
		defaultConfig: cconf.NewConfigParamsFromTuples(
			"options.retry_connect", true,
			"options.connect_timeout", 0,
			"options.reconnect_timeout", 3000,
			"options.max_reconnect", 3,
			"options.flush_timeout", 3000,
		),

		Logger:             clog.NewCompositeLogger(),
		ConnectionResolver: NewNatsConnectionResolver(),
		Options:            cconf.NewEmptyConfigParams(),

		subscriptions: []*NatsSubscription{},

		retryConnect:     true,
		maxReconnect:     3,
		reconnectTimeout: 3000,
		flushTimeout:     3000,
	}
	return c
}

// Configures component by passing configuration parameters.
//   - config    configuration parameters to be set.
func (c *NatsConnection) Configure(config *cconf.ConfigParams) {
	config = config.SetDefaults(c.defaultConfig)
	c.ConnectionResolver.Configure(config)

	c.Options = c.Options.Override(config.GetSection("options"))

	c.retryConnect = config.GetAsBooleanWithDefault("options.retry_connect", c.retryConnect)
	c.maxReconnect = config.GetAsIntegerWithDefault("options.max_reconnect", c.maxReconnect)
	c.reconnectTimeout = config.GetAsIntegerWithDefault("options.reconnect_timeout", c.reconnectTimeout)
	c.flushTimeout = config.GetAsIntegerWithDefault("options.flush_timeout", c.flushTimeout)
}

// Sets references to dependent components.
//   - references 	references to locate the component dependencies.
func (c *NatsConnection) SetReferences(references cref.IReferences) {
	c.Logger.SetReferences(references)
	c.ConnectionResolver.SetReferences(references)
}

// Checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *NatsConnection) IsOpen() bool {
	return c.Connection != nil
}

// Opens the component.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   - Return 			error or nil no errors occured.
func (c *NatsConnection) Open(correlationId string) error {
	options, err := c.ConnectionResolver.Resolve(correlationId)
	if err != nil {
		return err
	}

	uri := options.GetAsString("uri")

	connectOptions := []nats.Option{
		// nats.RetryOnFailedConnect(c.retryConnect),
		nats.MaxReconnects(c.maxReconnect),
		nats.ReconnectWait(time.Millisecond * time.Duration(c.reconnectTimeout)),
	}

	username := options.GetAsString("username")
	password := options.GetAsString("password")
	if username != "" {
		connectOptions = append(connectOptions, nats.UserInfo(username, password))
	}

	token := options.GetAsString("token")
	if token != "" {
		connectOptions = append(connectOptions, nats.Token(token))
	}

	connection, err := nats.Connect(uri, connectOptions...)
	if err != nil {
		c.Logger.Error(correlationId, err, "Failed to connect to NATS server at "+uri)
		return err
	}

	c.Connection = connection

	c.Logger.Debug(correlationId, "Connected to NATS server at "+uri)

	return nil
}

// Closes component and frees used resources.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Return			 error or nil no errors occured
func (c *NatsConnection) Close(correlationId string) error {
	if c.Connection == nil {
		return nil
	}
	c.Connection.Close()
	c.Logger.Debug(correlationId, "Disconnected to NATS server")

	c.Connection = nil
	c.subscriptions = []*NatsSubscription{}

	return nil
}

func (c *NatsConnection) GetConnection() *nats.Conn {
	return c.Connection
}

func (c *NatsConnection) ReadQueueNames() ([]string, error) {
	return []string{}, nil
}

func (c *NatsConnection) CreateQueue() error {
	return nil
}

func (c *NatsConnection) DeleteQueue() error {
	return nil
}

func (c *NatsConnection) checkOpen() error {
	if c.Connection != nil {
		return nil
	}

	return cerr.NewInvalidStateError(
		"",
		"NOT_OPEN",
		"Connection was not opened",
	)
}

// Publish a message to a specified topic
//
// Parameters:
//  - subject a subject (topic) name
//  - message a message to be published
// Returns: error or nil for success
func (c *NatsConnection) Publish(subject string, message *nats.Msg) error {
	// Check for open connection
	err := c.checkOpen()
	if err != nil {
		return err
	}

	if subject != "" {
		message.Subject = subject
	}
	return c.Connection.PublishMsg(message)
}

// Subscribe to a topic
//
// Parameters:
//   - subject a subject (topic) name
//   - queue (optional) a queue group
//   - listener a message listener
// Returns: err or nil for success
func (c *NatsConnection) Subscribe(subject string, queue string, listener INatsMessageListener) error {
	// Check for open connection
	err := c.checkOpen()
	if err != nil {
		return err
	}

	// Create the subscription
	subscription := &NatsSubscription{
		Subject:    subject,
		QueueGroup: queue,
		Listener:   listener,
	}

	// Subscribe to topic
	if queue == "" {
		subscription.Handler, err = c.Connection.QueueSubscribe(subject, queue, listener.OnMessage)
	} else {
		subscription.Handler, err = c.Connection.Subscribe(subject, listener.OnMessage)
	}
	if err != nil {
		return err
	}

	// Add the subscription
	c.subscriptions = append(c.subscriptions, subscription)
	return nil
}

// Unsubscribe from a previously subscribed topic topic
//
// Parameters:
//   - subject a subject (topic) name
//   - queue (optional) a queue group
//   - listener a message listener
// Returns: err or nil for success
func (c *NatsConnection) Unsubscribe(subject string, queue string, listener INatsMessageListener) error {
	// Remove the subscription
	var removedSubscription *NatsSubscription
	for index, subscription := range c.subscriptions {
		if subscription.Subject == subject && subscription.QueueGroup == queue && subscription.Listener == listener {
			removedSubscription = subscription
			c.subscriptions = append(c.subscriptions[:index], c.subscriptions[index+1:]...)
			break
		}
	}

	// If nothing to remove then skip
	if removedSubscription == nil {
		return nil
	}

	// Unsubscribe from the topic
	if removedSubscription.Handler != nil {
		return removedSubscription.Handler.Unsubscribe()
	}

	return nil
}


queues/NatsMessageQueue.go
package queues

import (
	"fmt"
	"time"

	"github.com/nats-io/nats.go"
	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cqueues "github.com/pip-services3-go/pip-services3-messaging-go/queues"
)

/*
NatsMessageQueue are message queue that sends and receives messages via NATS message broker.

 Configuration parameters:

- subject:                       name of NATS topic (subject) to subscribe
- queue_group:                   name of NATS queue group
- connection(s):
  - discovery_key:               (optional) a key to retrieve the connection from  IDiscovery
  - host:                        host name or IP address
  - port:                        port number
  - uri:                         resource URI or connection string with all parameters in it
- credential(s):
  - store_key:                   (optional) a key to retrieve the credentials from  ICredentialStore
  - username:                    user name
  - password:                    user password
- options:
  - serialize_message:    (optional) true to serialize entire message as JSON, false to send only message payload (default: true)
  - autosubscribe:        (optional) true to automatically subscribe on option (default: false)
  - retry_connect:        (optional) turns on/off automated reconnect when connection is log (default: true)
  - max_reconnect:        (optional) maximum reconnection attempts (default: 3)
  - reconnect_timeout:    (optional) number of milliseconds to wait on each reconnection attempt (default: 3000)
  - flush_timeout:        (optional) number of milliseconds to wait on flushing messages (default: 3000)


 References:

- *:logger:*:*:1.0             (optional)  ILogger components to pass log messages
- *:counters:*:*:1.0           (optional)  ICounters components to pass collected measurements
- *:discovery:*:*:1.0          (optional)  IDiscovery services to resolve connections
- *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials
- *:connection:nats:*:1.0      (optional) Shared connection to NATS service

See MessageQueue
See MessagingCapabilities

Example:

    queue := NewNatsMessageQueue("myqueue")
    queue.Configure(cconf.NewConfigParamsFromTuples(
      "subject", "mytopic",
	  "queue_group", "mygroup",
      "connection.protocol", "nats"
      "connection.host", "localhost"
      "connection.port", 1883
    ))

    queue.open("123")

    queue.Send("123", NewMessageEnvelope("", "mymessage", "ABC"))

    message, err := queue.Receive("123")
	if (message != nil) {
		...
		queue.Complete("123", message);
	}
*/
type NatsMessageQueue struct {
	*NatsAbstractMessageQueue

	autoSubscribe bool
	subscribed    bool
	messages      []cqueues.MessageEnvelope
	receiver      cqueues.IMessageReceiver
}

// NewNatsMessageQueue are creates a new instance of the message queue.
// Parameters:
//   - name  string (optional) a queue name.
func NewNatsMessageQueue(name string) *NatsMessageQueue {
	c := NatsMessageQueue{}

	c.NatsAbstractMessageQueue = InheritNatsAbstractMessageQueue(&c, name,
		cqueues.NewMessagingCapabilities(false, true, true, true, true, false, false, false, true))

	c.messages = make([]cqueues.MessageEnvelope, 0)

	return &c
}

// Configures component by passing configuration parameters.
//   - config    configuration parameters to be set.
func (c *NatsMessageQueue) Configure(config *cconf.ConfigParams) {
	c.NatsAbstractMessageQueue.Configure(config)

	c.autoSubscribe = config.GetAsBooleanWithDefault("options.autosubscribe", c.autoSubscribe)
}

// Opens the component with given connection and credential parameters.
// Parameters:
//   - correlationId     (optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Open(correlationId string) error {
	if c.IsOpen() {
		return nil
	}

	err := c.NatsAbstractMessageQueue.Open(correlationId)
	if err != nil {
		return err
	}

	// Subscribe right away
	if c.autoSubscribe {
		err = c.subscribe(correlationId)
		if err != nil {
			c.Close(correlationId)
			return err
		}
	}

	return nil
}

// Close method are Closes component and frees used resources.
// Parameters:
//   - correlationId string 	(optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Close(correlationId string) error {
	if !c.IsOpen() {
		return nil
	}

	err := c.NatsAbstractMessageQueue.Close(correlationId)

	// Unsubscribe from topic
	if c.subscribed {
		subject := c.SubscriptionSubject()
		c.Connection.Unsubscribe(subject, c.QueueGroup, c)
		c.subscribed = false
	}

	c.Lock.Lock()
	defer c.Lock.Unlock()
	c.receiver = nil
	c.messages = make([]cqueues.MessageEnvelope, 0)

	return err
}

func (c *NatsMessageQueue) subscribe(correlationId string) error {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	// Check if already were subscribed
	if c.subscribed {
		return nil
	}

	// Subscribe to the topic
	subject := c.SubscriptionSubject()
	err := c.Connection.Subscribe(subject, c.QueueGroup, c)
	if err != nil {
		c.Logger.Error(correlationId, err, "Failed to subscribe to subject "+subject)
		return err
	}

	c.subscribed = true
	return nil
}

// Clear method are clears component state.
// Parameters:
//   - correlationId 	string (optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Clear(correlationId string) (err error) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	c.messages = make([]cqueues.MessageEnvelope, 0)
	c.receiver = nil

	return nil
}

// ReadMessageCount method are reads the current number of messages in the queue to be delivered.
// Returns number of messages or error.
func (c *NatsMessageQueue) ReadMessageCount() (count int64, err error) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	count = (int64)(len(c.messages))
	return count, nil
}

// Peek method are peeks a single incoming message from the queue without removing it.
// If there are no messages available in the queue it returns nil.
// Parameters:
//   - correlationId  string  (optional) transaction id to trace execution through call chain.
// Returns: result *cqueues.MessageEnvelope, err error
// message or error.
func (c *NatsMessageQueue) Peek(correlationId string) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(correlationId)
	if err != nil {
		return nil, err
	}

	var message *cqueues.MessageEnvelope

	// Pick a message
	c.Lock.Lock()
	if len(c.messages) > 0 {
		message = &c.messages[0]
	}
	c.Lock.Unlock()

	if message != nil {
		c.Logger.Trace(message.CorrelationId, "Peeked message %s on %s", message, c.String())
	}

	return message, nil
}

// PeekBatch method are peeks multiple incoming messages from the queue without removing them.
// If there are no messages available in the queue it returns an empty list.
// Important: This method is not supported by NATS.
// Parameters:
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - messageCount      a maximum number of messages to peek.
// Returns:          callback function that receives a list with messages or error.
func (c *NatsMessageQueue) PeekBatch(correlationId string, messageCount int64) ([]*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(correlationId)
	if err != nil {
		return nil, err
	}

	c.Lock.Lock()
	batchMessages := c.messages
	if messageCount <= (int64)(len(batchMessages)) {
		batchMessages = batchMessages[0:messageCount]
	}
	c.Lock.Unlock()

	messages := []*cqueues.MessageEnvelope{}
	for _, message := range batchMessages {
		messages = append(messages, &message)
	}

	c.Logger.Trace(correlationId, "Peeked %d messages on %s", len(messages), c.Name())

	return messages, nil
}

// Receive method are receives an incoming message and removes it from the queue.
// Parameters:
//  - correlationId   string   (optional) transaction id to trace execution through call chain.
//  - waitTimeout  time.Duration     a timeout in milliseconds to wait for a message to come.
// Returns:  result *cqueues.MessageEnvelope, err error
// receives a message or error.
func (c *NatsMessageQueue) Receive(correlationId string, waitTimeout time.Duration) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(correlationId)
	if err != nil {
		return nil, err
	}

	messageReceived := false
	var message *cqueues.MessageEnvelope
	elapsedTime := time.Duration(0)

	for elapsedTime < waitTimeout && !messageReceived {
		c.Lock.Lock()
		if len(c.messages) == 0 {
			c.Lock.Unlock()
			time.Sleep(time.Duration(100) * time.Millisecond)
			elapsedTime += time.Duration(100)
			continue
		}

		// Get message from the queue
		message = &c.messages[0]
		c.messages = c.messages[1:]

		// Add messages to locked messages list
		messageReceived = true
		c.Lock.Unlock()
	}

	return message, nil
}

func (c *NatsMessageQueue) OnMessage(msg *nats.Msg) {
	// Deserialize message
	message, err := c.ToMessage(msg)
	if err != nil {
		c.Logger.Error("", err, "Failed to read received message")
	}

	c.Counters.IncrementOne("queue." + c.Name() + ".received_messages")
	c.Logger.Debug(message.CorrelationId, "Received message %s via %s", msg, c.Name())

	// Send message to receiver if its set or put it into the queue
	c.Lock.Lock()
	if c.receiver != nil {
		receiver := c.receiver
		c.Lock.Unlock()
		c.sendMessageToReceiver(receiver, message)
	} else {
		c.messages = append(c.messages, *message)
		c.Lock.Unlock()
	}
}

func (c *NatsMessageQueue) sendMessageToReceiver(receiver cqueues.IMessageReceiver, message *cqueues.MessageEnvelope) {
	correlationId := message.CorrelationId

	defer func() {
		if r := recover(); r != nil {
			err := fmt.Sprintf("%v", r)
			c.Logger.Error(correlationId, nil, "Failed to process the message - "+err)
		}
	}()

	err := receiver.ReceiveMessage(message, c)
	if err != nil {
		c.Logger.Error(correlationId, err, "Failed to process the message")
	}
}

// Listens for incoming messages and blocks the current thread until queue is closed.
// Parameters:
//  - correlationId   string  (optional) transaction id to trace execution through call chain.
//  - receiver    cqueues.IMessageReceiver      a receiver to receive incoming messages.
//
// See IMessageReceiver
// See receive
func (c *NatsMessageQueue) Listen(correlationId string, receiver cqueues.IMessageReceiver) error {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return err
	}

	// Subscribe if needed
	err = c.subscribe(correlationId)
	if err != nil {
		return err
	}

	c.Logger.Trace("", "Started listening messages at %s", c.Name())

	// Get all collected messages
	c.Lock.Lock()
	batchMessages := c.messages
	c.messages = []cqueues.MessageEnvelope{}
	c.Lock.Unlock()

	// Resend collected messages to receiver
	for _, message := range batchMessages {
		receiver.ReceiveMessage(&message, c)
	}

	// Set the receiver
	c.Lock.Lock()
	c.receiver = receiver
	c.Lock.Unlock()

	return nil
}

// EndListen method are ends listening for incoming messages.
// When this method is call listen unblocks the thread and execution continues.
// Parameters:
//   - correlationId  string   (optional) transaction id to trace execution through call chain.
func (c *NatsMessageQueue) EndListen(correlationId string) {
	c.Lock.Lock()
	c.receiver = nil
	c.Lock.Unlock()
}
